src/helpers/ring.rs

```diff
@@ -26,183 +27,270 @@ pub enum HelperAddr {
+    /// started by it and routes messages based on protocol identifiers they carry on with them.
+    #[derive(Debug)]
+    pub struct TestHelper {
+        left: Sender<MessageEnvelope>,
+        right: Sender<MessageEnvelope>,
+        routing_table: Arc<Mutex<RoutingTable>>,
+    }
+
+    /// Communication channel for test helpers. Holds references to channel senders to send messages
+    /// to the outside and owns the receivers (one per peer) to accept messages sent to it.
+    /// Owner control the lifetime of protocol communication links:
+    /// when this instance is dropped, link is removed.
+    #[derive(Debug)]
+    pub struct TestChannel {
+        protocol_id: ProtocolId,
+        table: Arc<Mutex<RoutingTable>>,
+        senders: HashMap<HelperAddr, Sender<MessageEnvelope>>,
+        receivers: HashMap<HelperAddr, Receiver<MessageEnvelope>>,
+    }
 
-    /// Each message is packed inside an envelope with some meta information about it.
     #[derive(Debug)]
     struct MessageEnvelope {
         source: HelperAddr,
-        type_id: TypeId,
+        protocol_id: ProtocolId,
         payload: Box<[u8]>,
     }
 
-    /// A mock implementation of `Ring` trait to be used in unit tests where all helpers are running
-    /// inside the same process. Provides simple and inefficient implementation by buffering all messages
-    /// on `send` and polling the buffer on `receive`. Message is determined to be the same if it has
-    /// the same `TypeId` and helper address matches the destination. For example, message `Foo` sent
-    /// to Helper 2 will be received and removed from the local buffer only when Helper 2 attempts to
-    /// receive it.
+    /// Communication infrastructure required for a given protocol. Senders represent outbound
+    /// links provided to helpers to send messages, receivers allow helpers to receive messages.
     #[derive(Debug)]
-    pub struct TestHelper {
-        // A handle to send message to this helper
-        input_queue: Sender<MessageEnvelope>,
-
-        // Reference to helper channel on the left side
-        left: Option<Sender<MessageEnvelope>>,
+    struct Route {
+        outbound: HashMap<HelperAddr, Sender<MessageEnvelope>>,
 
-        // Reference to helper channel on the right side
-        right: Option<Sender<MessageEnvelope>>,
+        /// Receivers for messages sent to this helper within the same protocol id. Helper takes
+        /// ownership of these receivers when connected and replaces the value with `None`.
+        /// is set to `None` once test helper is connected to this instance
+        inbound: Option<HashMap<HelperAddr, Receiver<MessageEnvelope>>>,
+    }
 
-        // buffer for messages sent to this helper
-        buf: Arc<Mutex<MessageBuf>>,
+    /// Keeps track of active protocols
+    #[derive(Debug, Default)]
+    struct RoutingTable {
+        active_routes: HashMap<ProtocolId, Route>,
```

Comment:
Ideally, the protocol ID is a very small number, with this being a `Vec<Route>` or similar.

If we assume that the protocol ID corresponds to a step in the overall process, we could make this a EnumMap and the protocol ID could be a short enum.  The problem is that you would need to make this type generic (which flows on upstream).

Reply:
the std implementation of hashmap is pretty good imo. It uses linear probing so it is cache friendly and allow sparse keys which is a clear advantage over pure `Vec`. it is also more ergonomic to use as I don't need to worry about growth